### DISCUSSION TITLE
how `this.resolve()` resolves file extension?

### DISCUSSION THREAD

**Discussion by centraldogma99** [rollup/rollup#5673] [Q&A] [answered]
I am currently working on a React project using Vite(which is build on top of rollup). When importing modules, I usually omit the file extensions, like import { Some } from '#libs/some' and worked with no problem. However, I recently discovered that Vite(More specifically, vite:import-analysis) was automatically adding extensions to import statements, But couldn’t find any specific documentation about this behavior in the Vite official docs.
<img width="1006" alt="스크린샷 2024-09-22 오후 2 37 12" src="https://github.com/user-attachments/assets/9dc7867b-b15e-4f68-ac29-f82afd1048a6">

So I analyzed Vite’s source code and discovered that it utilizes Rollup’s `this.resolve()` to achieve the automatic addition of file extensions. My question has now shifted to how Rollup’s this.resolve() determines the file extension, but I have been unable to find an answer.

Could you point me to any documentation or source code that might help resolve this question?

**Comment by lukastaegert** [comment, marked as answer]
`this.resolve` first hands over to all plugins that implement `resolveId`. In Vite, this would include `@rollup/plugin-node-resolve`, and this would be the one responsible for adding the extension. It does so by utilizing https://www.npmjs.com/package/resolve which again I think does a lookup in the local directory.
In the absence of any plugins, Rollup will still allow to omit certain extensions, which is handled here https://github.com/rollup/rollup/blob/bc7780c322e134492f40a76bf64afe561670425c/src/utils/resolveId.ts#L61-L71
